pgxn/neon/libpagestore.c

```diff
@@ -527,37 +699,48 @@ pageserver_disconnect(shardno_t shard_no)
 static void
 pageserver_disconnect_shard(shardno_t shard_no)
 {
+	PageServer *shard = &page_servers[shard_no];
 	/*
 	 * If anything goes wrong while we were sending a request, it's not clear
 	 * what state the connection is in. For example, if we sent the request
 	 * but didn't receive a response yet, we might receive the response some
 	 * time later after we have already sent a new unrelated request. Close
 	 * the connection to avoid getting confused.
 	 */
-	if (page_servers[shard_no].conn)
+	if (shard->state != PS_Disconnected)
 	{
-		neon_shard_log(shard_no, LOG, "dropping connection to page server due to error");
-		PQfinish(page_servers[shard_no].conn);
-		page_servers[shard_no].conn = NULL;
+		Assert(shard->conn);
+		Assert(shard->wes_read);
+
+		FreeWaitEventSet(shard->wes_read);
+		FreeWaitEventSet(shard->wes_write);
+		PQfinish(shard->conn);
+
+		shard->wes_read = NULL;
+		shard->wes_write = NULL;
+		shard->conn = NULL;
+		shard->state = PS_Disconnected;
 	}
-	if (page_servers[shard_no].wes != NULL)
+	else
 	{
-		FreeWaitEventSet(page_servers[shard_no].wes);
-		page_servers[shard_no].wes = NULL;
+		Assert(!shard->conn);
+		Assert(!shard->wes_read);
```

Comment:
If the CreateWaitEventSet() or AddWaitEventToSet() calls in pageserver_connect() fail, e.g. because of "out of memory", you might well end up with `shard->state == PS_Disconnected && shard->conn != NULL`.